sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -160,6 +168,16 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
   private def tryParseDouble(field: String): DataType = {
     if ((allCatch opt field.toDouble).isDefined || isInfOrNan(field)) {
       DoubleType
+    } else {
+      tryParseDateFormat(field)
+    }
+  }
+
+  private def tryParseDateFormat(field: String): DataType = {
+    if (options.inferDateType
+      && !dateFormatter.isInstanceOf[LegacySimpleDateFormatter]
```

Comment:
I didn't get the check. Could you explain, please, why do you avoid the formatter?

Reply:
It has to be LegacyFastDateFormatter, missed to changed it. Previously I was using the SimpleDateFormatter so added this LegacySimpleDateFormatter, Now since we are using the FastDateFormatter it has to be LegacyFastDateFormatter, Making that change.

If legacy is on, we have ambiguity about Datetype pattern matching, because they can be arbitrarily set by users.
It does not do the exact match, which means it's not going to distinguish yyyy-MM and yyyy-MM-dd for input, for instance, 2010-10-10.